proposed/2022/pack-referenced-projects.md

```diff
@@ -0,0 +1,64 @@
+- GitHub PR (-)
+
+## Summary
+
+Allow referenced projects' artifacts to be packed inside the NuGet.
+
+## Motivation
+
+When SDK-style project is being packed as a NuGet, all projects that it references are considered as NuGet themselves that this NuGet depends on.
+This is not always a desired behavior, sometimes user needs artifacts from referenced project(s) to be packed into the NuGet.
+Currently there's no easy way to override this behavior except for ditching `csproj`, and  going back to `nuspec`, which becomes another source of truth and all that good stuff.
+
+## Explanation
+
+### Functional explanation
+
+Proposed solution is to add new `Pack` property to `ProjectReference` (`<ProjectReference Include="..." Pack="True" />`) and `PackageReference`. If it's `False` or not present - consider the project (or package) as a NuGet's dependency, if it's `True` - pack its artifacts in.
+
+For `PackageReference`s it's useful in case referenced package is from private feed, but NuGet that's being packed will be public.
+
+### Technical explanation
+
+When NuGet is being created from `csproj` file:
+
+1) If `ProjectReference` doesn't have `Pack` property, or has it set to `False` - preserve behavior that have existed before introduction of this feature (don't include artifacts of referenced projects into the NuGet, and consider them as NuGet-dependencies).
+2) If `ProjectReference` has `Pack` property and it's set to `True` - include it's artifacts into the NuGet.
+3) Do the same steps for `PackageReference`s
+
+## Drawbacks
+
+None
+
+## Rationale and alternatives
+
+- Use `IsPackable` property from referenced project to determine it it's supposed to be packed or be a NuGet dependency.
+
+It will be more automated, but won't fit some edge cases (also it will be a breaking change). And it isn't applicable to `PackageReferences`.
+
+- Introduce new project-wide `PackReferencedProjects` property that would switch pack behavior for all `ProjectReference`s inside a single project.
+
+This one is less flexible since it cannot switch on per-`ProjectReference` basis. And it isn't applicable to `PackageReferences`.
+
+- Another solution is to introduce a `IncludeReferencedProjects` switch for `nuget` CLI.
+
+Downsides are: Behavior of individual projects (if packing is performed in bulk) and `ProjectReference`s cannot be changed, and it's not obvious how to properly pack any specific project (I thought we are consolidating all the information in `csproj` now). And it isn't applicable to `PackageReferences`.
+
+## Prior Art
+
+This design flaw have been a pain for the community for more than 5 years now. It was such an issue that even third party solution was created - [nugetizer](https://github.com/devlooped/nugetizer).
+
+I haven't used it, so don't have much to comment here, but community clearly [expressed a desire for out of the box solution](https://github.com/NuGet/Home/issues/3891#issuecomment-1080044314).
+
+## Unresolved Questions
+
+None
```

Comment:
One of the primary use cases of bundling third party package content within your own package is custom build logic - analyzers, custom msbuild tasks, bundling satellite applications. There will be certain pressure to not only resolve a single level of package but a xcopy ready representation of that dependency's own dependency tree. To give some examples:

- a custom build task i am working on at the moment uses `Microsoft.EntityFrameworkCore.Sqlite` to write out gathered information as a database file
  - that package incurs a slew of packages dependent on TFM & platform - Do we stop after bundling the upper node of the dependency tree specified?
- an analyzer package needs a parser library at runtime - where does that `ProprietaryFileFormatParserXLib` land without further instructions - `PackagePath` anyone?

A possible basic implementation of the feature would look like a 1:1 copy of `PackageDependencyA/lib/[TFM or any]/` to `PackageToBePacked/lib/[TFM or any]/` - this would mean that any library package added into a bundled package would end up being referenced by consumers of that bundled package - is that a desired outcome?

I would recommend adding the following questions:

```suggestion
- **How do we flow/assign TFM based `lib`/`content` items?** (Is `PackagePath` a desired parameter?)
- **How is conflict resolution going to work with consumers taking in a dependency already present in the bundle?**
- Should license flow/validation be considered right from the start?
- **Should a `PackageReference` or `ProjectReference` tagged with `Pack="true"` be resolved transitively?**
```

Reply:
cc: @AraHaan

Reply:
1. I think usually one can check that as an implementation detail using:
- ``$(TargetFramework)``

However there comes another task it would have to do as well:
- use the value of ``$(RuntimeIdentifier)`` if set in project (otherwise ignore it when it == ``any``).

2. I think if analyzers/generators are referenced in project and are tagged with ``Pack="true"`` that it should package everything under the analyzer's folder in it's package AS IS which means conflicts between internal deps of the analyzers could possibly happen.

3. As for the license flow, what it could do is add these to the package metadata to support it: ``licenses/<package name>`` where it would bundle the licenses from the included projects/packages.
4. That just depends entirely on a lot of factors, if the package is meant only for an generator / analyzer (nothing more) than I say that resolving them would not be needed, otherwise yes they would be needed (we do not want to resolve dependencies of analyzers that would never be used on the consuming project and so wastes disc space).